Analysis/R/prepare_stan_input.R

```diff
@@ -384,7 +383,290 @@ prepare_stan_input <- function(
   # What observation model to use
   stan_data$obs_model <- config$obs_model
   
-  # ---- J. Data for output summaries ----
+  # First define admin levels to get data ad upper admin levels
+  # Administrative levels for observation model
+  admin_levels <- sf_cases_resized$location_name %>% 
+    taxdat::get_admin_level() %>% 
+    as.array()
+  
+  n_na_admin <- sum(is.na(admin_levels))
+  
+  if (n_na_admin > 0) {
+    cat("---- Replacing unknown admin level for ", n_na_admin, " observations corresponding to",
+        sum(stan_data$y[is.na(admin_levels)]), "cases. \n")
+  }
+  
+  # Make sure all admin levels are specified
+  admin_levels[is.na(admin_levels)] <- max(admin_levels, na.rm = T)
+  
+  # Get unique levels, this is necessary if not all admin levels are present
+  # in the data
+  u_admin_levels <- sort(unique(admin_levels))
+  
+  # index staring at 1
+  stan_data$map_obs_admin_lev <- purrr::map_dbl(admin_levels, ~ which(u_admin_levels == .))
+  
+  # Add unique number of admin levels for use in observation model
+  stan_data$N_admin_lev <- length(u_admin_levels)
+  
+  # Map from space x time grid to space grid
+  sf_grid <- sf_grid %>% 
+    dplyr::group_by(rid, x, y) %>% 
+    dplyr::mutate(space_id = min(upd_id)) %>% 
+    dplyr::ungroup()
+  
+  stan_data$N_space <- length(unique(sf_grid$space_id))
+  stan_data$map_spacetime_space_grid <- sf_grid$space_id[sf_grid$upd_id]
+  
+  # ---- I. Mean rate ----
+  stan_data$meanrate <- taxdat::compute_mean_rate(stan_data = stan_data)
+  
+  #  ---- J. Imputation ----
```

Comment:
I would recommend moving the imputation into functions so we can do testing... Even if you don't write the unit tests now, we can have @fang19911030 help us do that later

Reply:
Ok I will try put into functions.